File: cpp/creational/registry/nishith/Registry/Registry/src/DefaultInternsInitializer.cpp
#include "DefaultInternsInitializer.h"
#include "InternKeys.h"
#include "BatchInfo.h"
#include "BusinessAnalystIntern.h"
#include "EngineeringIntern.h"

using namespace std;

void register_default_interns(InternRegistry& registry) {
  // Create Batch for Engineering Interns
  const auto eng_batch = std::make_shared<BatchInfo>();
  eng_batch->name = "2025-ENG";
  eng_batch->start_date = "2025-06-01";
  eng_batch->end_date = "2025-12-01";
  eng_batch->duration = "6 months";
  eng_batch->location = "Bangalore";

  auto engineering_intern = std::make_unique<EngineeringIntern>();
  engineering_intern->set_batch_info(eng_batch);
  registry.Register(intern_key::engineering, std::move(engineering_intern));

  // Create Batch for Business Analyst Interns
  const auto ba_batch = std::make_shared<BatchInfo>();
  ba_batch->name = "2025-BA";
  ba_batch->start_date = "2025-07-01";
  ba_batch->end_date = "2025-11-30";
  ba_batch->duration = "5 months";
  ba_batch->location = "Hyderabad";

  auto new_ba_intern = std::make_unique<BusinessAnalystIntern>();
  new_ba_intern->set_batch_info(ba_batch);
  registry.Register(intern_key::business_analyst, std::move(new_ba_intern));
}

File: java/creational/registry/uday/DefaultInternsInitializer.java
/**
 * Utility class for registering default intern prototypes in the intern
 * registry. This class initializes predefined batches and associates them
 * with their corresponding intern types.
 */
public class DefaultInternsInitializer {

    /**
     * Registers default intern prototypes into the provided intern registry.
     * This includes prototypes for Engineering and Business Analyst interns
     * with predefined batch information.
     *
     * @param registry the {@code InternRegistry} to populate with default
     *                 intern prototypes.
     */
    public static void registerDefaultInterns(InternRegistry registry) {
        // Create Batch for Engineering Interns
        BatchInfo engBatch = new BatchInfo();
        engBatch.setName("2025-ENG");
        engBatch.setDuration("6 months");
        engBatch.setLocation("Bangalore");

        EngineeringIntern engineeringIntern = new EngineeringIntern();
        engineeringIntern.setBatchInfo(engBatch);
        registry.register(InternKeys.ENGINEERING, engineeringIntern);

        // Create Batch for Business Analyst Interns
        BatchInfo baBatch = new BatchInfo();
        baBatch.setName("2025-BA");
        baBatch.setDuration("5 months");
        baBatch.setLocation("Hyderabad");

        BusinessAnalystIntern newBaIntern = new BusinessAnalystIntern();
        newBaIntern.setBatchInfo(baBatch);
        registry.register(InternKeys.BUSINESS_ANALYST, newBaIntern);
    }
}
